addons/sale/models/sale_order_line.py

```diff
@@ -989,6 +1056,24 @@ def _compute_product_uom_readonly(self):
 
     #=== CONSTRAINT METHODS ===#
 
+    @api.constrains('combo_item_id')
+    def _check_combo_item_id(self):
+        """ `combo_item_id` should never be set manually. This constraint mainly serves to avoid
+        programming errors.
+        """
+        for line in self:
+            linked_line = line._get_linked_line()
+            allowed_combo_items = linked_line.product_template_id.mapped('combo_ids.combo_item_ids')
```

Comment:
Before 13/14?, you had to use `mapped` whenever you accessed a field on a multi-record recordset, but since then, mapped is implicitly used when reading relational fields, and the flattening indeed works automatically.